### ISSUE TITLE
Question regarding the reproduction of results

### ISSUE THREAD

**Issue by Ryenum** [YeFD/RRAG#2]
Thank you for making this excellent work open source! Currently, I am attempting to reproduce the experimental results using LLaMA-2-7B on the NQ-10 dataset, but I have encountered a significant performance gap compared to the values reported in the paper.
Regarding the evaluation metrics and settings, I have two questions: 
1. Regarding the description of the evaluation metrics
The paper reports ACC and F1 scores, but the evaluation code seems to use best_subspan_em.Could you confirm whether best_subspan_em is the main metric based on which the "ACC" reported in the table is calculated? I want to ensure that I am comparing the correct metrics. 
2. Significant performance gap
I noticed a large difference in the scores.My result: best_subspan_em is approximately 0.13 (after training it is 0.48).The reported result: the value reported in the paper is 0.38 (after training it is 0.69).I am using the LLaMA-2-7B model. Could you please share some suggestions regarding potential issues? I suspect that I might have missed some specific configurations. For example: for NQ-10, is there a need for specific a few examples or prompt formats?

Any guidance or relevant links regarding the specific evaluation script used in this paper would be greatly appreciated! Thank you in advance.

**Comment by yefd** [comment]
Thanks for your interest in our work!

1. Yes, all baselines and R²AG use `best_subspan_em` rather than exact matching `em` for the reported ACC. See metric implementation.

https://github.com/yefd/RRAG/blob/61cfd1027c0101172fda701ec816bc18360b5bab/RRAG/utils/metrics.py#L38-L45

2. We did not use any additional specially designed prompts. For both RAG and R²AG, See

https://github.com/yefd/RRAG/blob/61cfd1027c0101172fda701ec816bc18360b5bab/RRAG/prompts/qa.prompt#L1-L7

For the performance gap, one possible issue is the chat template setting. For LLaMA-2-7B, `instruction_type` should be set to `instruction`, while for LLaMA-2-7B-Chat it should be set to `chat`. Please see relevant code
https://github.com/yefd/RRAG/blob/61cfd1027c0101172fda701ec816bc18360b5bab/runner.py#L126-L155

Another possible reason is the document ordering. You may try the setting where the relevant document is placed at the first position, eg. [NQ-10 with GT at 0](https://github.com/nelson-liu/lost-in-the-middle/blob/main/qa_data/10_total_documents/nq-open-10_total_documents_gold_at_0.jsonl.gz).

Hope this helps!

**Comment by Ryenum** [comment]
Thank you very much for your previous response. It was very helpful.
I have been working on reproducing the project this week. I ran the experiments multiple times and tried several optimization methods, such as lowering the learning rate, enabling beam search, and increasing the number of training epochs. While these attempts led to slight improvements, the performance metrics are still hovering between 0.4 and 0.5, which is significantly lower than the results reported in the paper.
I have a few specific questions and would appreciate your guidance:
1. Regarding the Evaluation Metric
Is the best_subspan_em metric used in the code strictly equivalent to the ACC (Accuracy) metric reported in the paper? If they are not equivalent, could you clarify how I should compare my reproduction results with the original paper's numbers?

2. Regarding the Dataset and Preprocessing
I suspect the lower performance might be related to the dataset preprocessing.

- Original NQ: I initially downloaded the raw NQ dataset and extracted features using the project's code. I suspect this process might have caused data misalignment (e.g., shuffling the order of ground truth answers), leading to lower accuracy.
- Pre-processed Dataset: I then tried using the pre-processed dataset you provided. However, the format seems different from what the current code expects. My manually processed data follows the question/answers/cbxs format (compatible with load_nq.py), but your provided dataset uses an input/context/passages format, which causes errors when running.
(See the error log below)

   <img width="417" height="39" alt="Image" src="https://github.com/user-attachments/assets/4d3347dd-4e88-4d51-ba06-691d894e8e17" />

Could you provide some advice on how to correctly use the pre-processed dataset or verify if my data alignment is correct?

3. Regarding the Environment Configuration
- My local environment differs slightly from the original setup due to hardware constraints (specifically, older versions of PyTorch were not compatible with my GPU).
My Setup: torch==2.7.0+cu128, transformers==4.40.2, etc. (See screenshot below).
- Code Modifications: I made some minor code adjustments to ensure compatibility with this newer environment.
(See my environment details below)

   <img width="383" height="173" alt="Image" src="https://github.com/user-attachments/assets/7bcfa174-e6b3-4ee6-9fae-76ba6e2844b7" />

Do you think these environmental differences and code adaptations could be the primary cause of the performance gap?

Thank you again for your time and help!

**Comment by yefd** [comment]
1. Yes, all reported results in our experiments use best_subspan_em to ensure fair comparison across baselines and R²AG. Sorry for the misleading definition of Acc in the paper.
2. The [NQ-10 with GT at 0](https://github.com/nelson-liu/lost-in-the-middle/blob/main/qa_data/10_total_documents/nq-open-10_total_documents_gold_at_0.jsonl.gz) file is not a directly pre-processed dataset. And it contains the `question` field. You need to process the NQ-10 data and organize it into the load_nq.py format.
https://github.com/yefd/RRAG/blob/61cfd1027c0101172fda701ec816bc18360b5bab/scripts/feature_extraction/extraction_nq.sh#L1-L5
3. Regarding the performance gap, I think this is more likely caused by the dataset/candidate set rather than the environment. 

- Your RAG baseline performance is also quite low, which means the NQ version used in your implementation differs in candidate passages and difficulty. 
- Since R²AG improves after training in your experiments, the trend is consistent with our expectation.
- To reproduce the paper's performance, please make sure to use the lost-in-the-middle NQ-10 version.

**Comment by Ryenum** [comment]
I re-checked the dataset I used locally and found that the correct answers are not always at the 0th position (Gold at 0).

I ran a simple script to inspect the data distribution, and here are the results:

<img width="728" height="310" alt="Image" src="https://github.com/user-attachments/assets/a1f68274-6c58-4b8f-8233-59d3a8e27c31" />

As shown, the positions of the gold answers are quite scattered, with only about 21% of them appearing at the first position. We suspect that this might be the main reason for the performance gap (perhaps the dataset version I used is not perfectly aligned with the specific experimental setup in the paper).

That said, we still find the R²AG scheme to be highly effective and well-designed.

Thank you very much for your time and help!

**Comment by yefd** [comment]
Thanks for the update, and glad to know our work is effective in your experiments.

In our paper, we used a retriever that had been trained on NQ dataset, which may lead to better retrieval quality. You may also consider trying fine-tuned BERT or more recent retrievers (such as BGE and Qwen-Embedding).

Another practical suggestion is to try more recent LLMs, such as Qwen3. They usually have stronger ability, and may be less sensitive to document ordering than LLaMA-2-7B.

Wish you all the best!
